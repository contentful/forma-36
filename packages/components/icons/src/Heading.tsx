import React, { Fragment } from 'react';
import { generateIcon } from '@contentful/f36-icon';

export const Heading = generateIcon({
  name: 'Heading',
  path: (
    <Fragment>
      <path d="M0 0h24v24H0V0z" fill="none" />
      <path d="M14.8 5v5.5H9.2V5H6.4v14h2.8v-6.1h5.6V19h2.8V5z" />
    </Fragment>
  ),
});
