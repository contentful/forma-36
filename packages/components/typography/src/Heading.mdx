---
title: 'Heading'
type: 'component'
status: 'stable'
slug: /components/heading/
github: 'https://github.com/contentful/forma-36/tree/master/packages/components/typography/src'
storybook: 'https://f36-storybook.contentful.com/?path=/story/typography-heading--basic'
typescript: ./Heading.tsx
---

`Heading` is intended to be used to define a section. Headings are very important for accessibility, as it helps to make copy more easily scannable and guides a user through your content.

Another accessibility consideration screen readers. Visually impaired people often use screen readers, which find sections in your copy based on your heading hierarchy.

### Import

```js static=true
import { Heading } from '@contentful/f36-components';
// or
import { Heading } from '@contentful/f36-typography';
```

## Code examples

```jsx
<Heading>I love Forma 36 design system</Heading>
```

You can override the default margin bottom by specifying it explicitly.

```jsx
<Flex flexDirection="column">
  <Heading marginBottom="none">I love Forma 36 design system.</Heading>
  <Heading marginTop="spacingS">I love Forma 36 design system.</Heading>
</Flex>
```

## Props

import { Props } from '@contentful/f36-docs-utils';

<Props of="Heading" />
