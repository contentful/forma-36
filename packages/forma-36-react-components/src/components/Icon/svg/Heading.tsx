import * as React from 'react';

function SvgHeading(props: React.SVGProps<SVGSVGElement>) {
  return (
    <svg width="1em" height="1em" viewBox="0 0 24 24" {...props}>
      <path d="M0 0h24v24H0V0z" fill="none" />
      <path d="M14.8 5v5.5H9.2V5H6.4v14h2.8v-6.1h5.6V19h2.8V5z" />
    </svg>
  );
}

export default SvgHeading;
