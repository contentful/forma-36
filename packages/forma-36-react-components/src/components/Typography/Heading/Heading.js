import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './Heading.css';

class Heading extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    element: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p']),
  };

  static defaultProps = {
    element: 'h1',
    extraClassNames: undefined,
    testId: 'cf-ui-heading',
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      element,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.Heading, extraClassNames);

    const Element = element;

    return (
      <Element className={classNames} data-test-id={testId} {...otherProps}>
        {children}
      </Element>
    );
  }
}

export default Heading;
