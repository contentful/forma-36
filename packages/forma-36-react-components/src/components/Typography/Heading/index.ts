export * from './Heading';
export { default } from './Heading';
