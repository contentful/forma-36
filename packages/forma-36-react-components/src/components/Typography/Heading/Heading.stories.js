import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, selectV2 } from '@storybook/addon-knobs/react';
import { host } from 'storybook-host';
import { withInfo } from '@storybook/addon-info';

import Heading from './Heading';

storiesOf('Components|Typography/Heading', module)
  .addDecorator(
    host({
      align: 'center middle',
      cropMarks: false,
    }),
  )
  .add(
    'default',
    withInfo()(() => (
      <Heading
        extraClassNames={text('Extra Class Names', '')}
        element={selectV2(
          'Element',
          ['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'p'],
          'h1',
        )}
      >
        {text('Content', 'Heading')}
      </Heading>
    )),
  );
