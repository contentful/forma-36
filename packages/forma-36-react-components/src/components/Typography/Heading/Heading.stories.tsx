import React from 'react';
import Heading, { HeadingProps } from './Heading';

export default {
  title: 'Typography/Heading',
  component: Heading,
  parameters: {
    propTypes: [Heading['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
  },
};

interface Args extends HeadingProps {
  headingText: string;
}

export const Basic = ({ headingText, ...args }: Args) => (
  <Heading {...args}>{headingText}</Heading>
);

Basic.args = {
  headingText: 'Heading',
};
