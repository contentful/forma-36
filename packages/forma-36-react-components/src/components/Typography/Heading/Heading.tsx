import React, { Component } from 'react';
import cn from 'classnames';
import styles from './Heading.css';

import { TypographyContext } from '../Typography/Typography';

export interface HeadingProps {
  extraClassNames?: string;
  children?: React.ReactNode;
  testId?: string;
  element: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p';
  style?: React.CSSProperties;
}

export class Heading extends Component<HeadingProps> {
  static defaultProps = {
    element: 'h1',
    testId: 'cf-ui-heading',
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      element,
      ...otherProps
    } = this.props;

    const classNames = cn(styles['Heading'], extraClassNames);

    const Element = element;

    return (
      <TypographyContext.Consumer>
        {value => {
          return (
            <Element
              className={cn(classNames, [
                value['heading'] &&
                  `f36-margin-bottom--${value['heading']['spacing']}`,
              ])}
              data-test-id={testId}
              {...otherProps}
            >
              {children}
            </Element>
          );
        }}
      </TypographyContext.Consumer>
    );
  }
}

export default Heading;
