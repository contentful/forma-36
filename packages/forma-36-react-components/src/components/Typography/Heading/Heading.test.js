import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import Heading from './Heading';

it('renders the component', () => {
  const output = shallow(<Heading>Heading</Heading>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <Heading extraClassNames="my-extra-class">Heading</Heading>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component h3', () => {
  const output = shallow(<Heading element="h3">Heading</Heading>);

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(<Heading>Heading</Heading>).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
