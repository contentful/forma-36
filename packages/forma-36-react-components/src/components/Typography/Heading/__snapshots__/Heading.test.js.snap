// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<h1
  className="Heading"
  data-test-id="cf-ui-heading"
>
  Heading
</h1>
`;

exports[`renders the component h3 1`] = `
<h3
  className="Heading"
  data-test-id="cf-ui-heading"
>
  Heading
</h3>
`;

exports[`renders the component with an additional class name 1`] = `
<h1
  className="Heading my-extra-class"
  data-test-id="cf-ui-heading"
>
  Heading
</h1>
`;
